### PR TITLE
Fix #32517 - Moving rhsm parser and importer to core

### PR DESCRIPTION
**Pull request by domitea**
This PR continues from https://github.com/theforeman/foreman/pull/8491 and adding rshm parser with importer to core. 

It lacks a test for impoter. However, moving test for importer means also means move `Host::SuscriptionFacet` model (https://github.com/Katello/katello/blob/master/test/models/rhsm_fact_importer_test.rb#L20) and I don't think it's a good idea.

### CODE REVIEW

**Review by ekohl — COMMENTED** [review]
Did you intend to include the Ansible parser here?

**Comment by domitea** [comment]
@ekohl Well, it's not necessary, but I'm using a initializer made during moving Ansible parser.

**Comment by ekohl** [comment]
For future reference, then I'd always mention it in the PR description and link to that PR.

**Comment by ares** [comment]
@domitea as discussed on IRC, it would be good to add a brand new test just for the importer logic. All the rest is covered by the parent class test. I like your decision the changes to avoid the dependency on Consumer and SubscriptionFacet. Once the Ansible part is done, we can continue with the RHSM facts.

**Comment by ekohl** [comment]
For future reference, you can also consider a common commit that only adds the entry point(s). That can generally speaking be merged relatively quickly and means the whole Ansible PR isn't blocking the other reviews. It allows a higher concurrency on the reviews.

File: test/unit/foreman_ansible/fact_parser_test.rb [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        returns(sample_mock)
+      sample_mock.expects(:first_or_create)
+    end
+  end
+
+  # Tests for Network parser
+  class NetworkFactParserTest < ActiveSupport::TestCase
+    setup do
+      @facts_parser = AnsibleFactParser.new(
+        HashWithIndifferentAccess.new(
```
Comment (lzap) [inline test/unit/foreman_ansible/fact_parser_test.rb]: Oh I think I understand this comes from a different commit :-)

Please add some integration tests - those which have a fixture file (JSON or YAML or whatever) and successfuly import a host.

**Comment by lzap** [comment]
I have a one major requirement common for all these PRs: I would like to see integration tests. Unit tests are great, however a fixture JSON file that can be easily captured usually via "facter" or "ansible" or "subscription-manager" commands are very easy to add and manage. This way we can be adding multiple versions, OS versions or even network configuraitons (bond, bridge, vlan, vlan on top of bond). This really pays off in Discovery and I think we should have this from the day one. Ideally if you create some subdirectories in test/fixtures/fact_parser (or wherever we have current facts) and dynamically load them all.

You can do it in a separate PR if you like just please commit to this, I belive this is very important.

**Comment by domitea** [comment]
I understand your concern and I know that testing is very important part of developing. I'm planning also refactor tests for parsers and importers together with refactoring of these components. That means I need firstly have all parsers and importers in one place for refactoring.

File: test/unit/foreman_ansible/fact_parser_test.rb (hunk 1 above)
Comment (domitea) [reply test/unit/foreman_ansible/fact_parser_test.rb]: Well, PR is based on commit of adding Ansible parser due to using own initializer defined in Ansible PR.

**Comment by lzap** [comment]
Can you rebase, CI had a temporary problem with git and also tests are failing.

**Review by ekohl — CHANGES_REQUESTED** [review]
Needs a rebase now.

**Review by ekohl — COMMENTED** [review]
I do wonder if we'll see problems with the namespace. We did see those in Katello (https://github.com/Katello/katello/commit/2289a5506831b3cf9f5873f6e71c01c47e7792c1) before. That does make me more cautious to just merge this now.

Have you already opened a PR to remove the fact parser from Katello and if so, can you link it so we can merge them together?

**Comment by domitea** [comment]
@ekohl yes, there is also PR with removal from Katello (https://github.com/Katello/katello/pull/9346/files) and It's updated according to Foreman version requirements.

**Comment by ekohl** [comment]
Thanks for that. I'd like an ACK on that PR. The test failures also look relevant.

File: app/services/katello/rhsm_fact_parser.rb [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  class RhsmFactParser < ::FactParser
+    def architecture
+      name = facts['lscpu.architecture'] || facts['uname.machine']
+      name = "x86_64" if name == "amd64"
+      name = "i386" if name == "i686"
+      Architecture.where(:name => name).first_or_create if name.present?
+    end
+
+    def model
+      if facts['virt::is_guest'] == "true"
```
Comment (ekohl) [inline app/services/katello/rhsm_fact_parser.rb:14]: I know this is a direct import, but I'd expect `virtual` to be normalized to a real bool and reuse it here:
```suggestion
      if virtual
```

**Comment by domitea** [comment]
Tests are green, so I think it's ready for merge @ekohl

File: app/services/katello/rhsm_fact_parser.rb [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      true
+    end
+
+    def get_facts_for_interface(interface)
+      {
+        'link' => true,
+        'macaddress' => get_rhsm_mac(interface),
+        'ipaddress' => get_rhsm_ip(interface),
+      }
+    end
```
Comment (ekohl) [inline app/services/katello/rhsm_fact_parser.rb:32]: Making a note so we don't forget: this is IPv4 only but it should really also parse the IPv6 fact. I'm not sure if you keep any notes for the refactor, but @ShimShtein ran into this today.

**Review by ekohl — APPROVED** [review]
I can't merge the Katello bits, but we should merge them together. @tbrisker could you assist here?

File: app/services/katello/rhsm_fact_parser.rb (hunk 3 above)
Comment (domitea) [reply app/services/katello/rhsm_fact_parser.rb:32]: Thanks for the Info. I will update commit accordingly.

Comment (ekohl) [reply app/services/katello/rhsm_fact_parser.rb:32]: FYI: in Katello it's also IPv4-only so it'd be fine (perhaps even better) to do it as a separate issue.

**Review by tbrisker — APPROVED** [review]
Thanks @domitea !

**Comment by lzap** [comment]
Thanks!

**Pull request merged**